src/github.com/matrix-org/dendrite/clientapi/routing/register.go

```diff
@@ -376,6 +422,62 @@ func isValidMacLogin(
 	return hmac.Equal(givenMac, expectedMAC), nil
 }
 
+const letterBytes = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ"
+const (
+	letterIdxBits = 6                    // 6 bits to represent a letter index
+	letterIdxMask = 1<<letterIdxBits - 1 // All 1-bits, as many as letterIdxBits
+	letterIdxMax  = 63 / letterIdxBits   // # of letter indices fitting in 63 bits
+)
+
+var src = rand.NewSource(time.Now().UnixNano())
+
+// RandString returns a random string of characters with a given length.
+// Do note that it is not thread-safe in its current form.
+// https://stackoverflow.com/a/31832326
+func RandString(n int) string {
+	b := make([]byte, n)
+
+	// A src.Int63() generates 63 random bits
+	for i, cache, remain := n-1, src.Int63(), letterIdxMax; i >= 0; {
+		if remain == 0 {
+			cache, remain = src.Int63(), letterIdxMax
+		}
+		if idx := int(cache & letterIdxMask); idx < len(letterBytes) {
+			b[i] = letterBytes[idx]
+			i--
+		}
+		cache >>= letterIdxBits
+		remain--
+	}
+
+	return string(b)
+}
+
+// checkFlowsEqual checks if two registration flows have the same stages
+// within them. Order of stages does not matter.
+func checkFlowsEqual(aFlow, bFlow authtypes.Flow) bool {
+	a := aFlow.Stages
+	b := bFlow.Stages
+	if len(a) != len(b) {
+		return false
+	}
+
+	aCopy := make([]string, len(a))
+	bCopy := make([]string, len(b))
+
+	for loginType := range a {
+		aCopy = append(aCopy, string(loginType))
+	}
+	for loginType := range b {
+		bCopy = append(bCopy, string(loginType))
+	}
+
+	sort.Strings(aCopy)
+	sort.Strings(bCopy)
+
+	return reflect.DeepEqual(aCopy, bCopy)
```

Comment:
Since these are just slices of strings I'd just do a for loop. I don't trust the reflect package not to be weird or slow.

```golang
for i := range aCopy {
    if a[i] != b[i] {
        return false
    }
}
return true
```

Reply:
It might actually be nicer to just do, instead of creating copies:

```golang
sort.Slice(aFlow, func(x, y) { return x < y })
```

Reply:
> It might actually be nicer to just do, instead of creating copies:
>     `sort.Slice(aFlow, func(x, y) { return x < y })`

Oh, much cleaner. I'll give that a whirl.

Reply:
Nice, works: https://play.golang.com/p/wLPjs2gfWE